File: courses/linear_regression/simulating_classic_stat/simulation_exercise.R
# Exercise 1, check if 
n <- 100
num_sim <- 1000
outcome <- rep(NA, num_sim)

for(i in 1:num_sim){
    new_data <- rnorm(n, 0, 5)
    y_avg <- mean(new_data)
    # In class I forgot the sqrt(n-1) factor!
    y_se <- sd(new_data) / sqrt(n)
    stat <- abs(y_avg / y_se)
    outcome[i] <- stat >= 1.96
}

mean(outcome)
# We should expect this to surround 
2 * pt(-1.96, n-1)
# If we did not estimate y_se, i.e. y_se = 5 / sqrt(n) then we would expect
2 * pnorm(-1.96)



# Exercise 2, is the median a better estimate than the avg for E(Y)
# When Y~Normal(0, 5^2)
bad_est <- function(x){
    if(sample(c(TRUE, FALSE), 1)){
        out <- quantile(x, 0.75)
    }else{
        out <- quantile(x, 0.25)
    }
    return(out)
}
n <- 100
num_sim <- 1000
outcome <- matrix(NA, nrow=num_sim, ncol=3)
true_mean = 0
for(i in 1:num_sim){
    new_data <- rnorm(n, true_mean, 5)
    outcome[i, 1] <- mean(new_data)
    outcome[i, 2] <- median(new_data)
    outcome[i, 3] <- bad_est(new_data)
}

x_range <- range(outcome)
par(mfrow=c(1, 3))
hist(outcome[, 1], xlim=x_range, main="avg")
hist(outcome[, 2], xlim=x_range, main="med")
hist(outcome[, 3], xlim=x_range, main="bad")

# Not recommended
abs(apply(outcome, 2, mean) - true_mean)
# Better
apply(outcome, 2, function(x) mean(abs(x - true_mean)))
apply(outcome, 2, function(x) mean((x - true_mean)^2))

File: courses/linear_regression/simulating_classic_stat/simulation_exercise.py
import random
import numpy as np
import matplotlib.pyplot as plt

# Exercise 1
n = 100
num_sim = 1000
outcome = []

for i in range(num_sim):
    new_data = np.random.normal(0, 5, n)
    y_avg = np.mean(new_data)
    # In class I forgot the sqrt(n) factor!
    y_sd = np.sqrt(sum((new_data - y_avg)**2) / (n-1))
    y_se = y_sd / np.sqrt(n)
    stat = abs(y_avg / y_se)
    outcome.append(stat >= 1.96)

np.mean(outcome)
# Notice this is a t-distribution with n-1 degrees of freedom


# Exercise 2, is the median a better estimate than the avg for E(Y)
# When Y~Normal(0, 5^2)

def bad_est(x):
    if random.choice([True, False]):
        out = np.percentile(x, 75)
    else:
        out = np.percentile(x, 25)
    return out


n = 100
num_sim = 1000
outcome = np.empty((num_sim, 3))
true_mean = 0
for i in range(num_sim):
    new_data = np.random.normal(true_mean, 5, n)
    outcome[i, 0] = np.mean(new_data)
    outcome[i, 1] = np.percentile(new_data, 50)
    outcome[i, 2] = bad_est(new_data)

fig, axes = plt.subplots(1, 3)
for i in range(outcome.shape[1]):
    axes[i].hist(outcome[:, i])
    axes[i].set_xlim(np.min(outcome), np.max(outcome))
plt.show()

# Not recommended
np.abs(np.apply_along_axis(np.mean, 0, outcome) - true_mean)
# Better
np.apply_along_axis(lambda x: np.mean(np.power(x - true_mean, 2)), 0, outcome)
np.apply_along_axis(lambda x: np.mean(np.abs(x - true_mean)), 0, outcome)
